src/main/scala/ch/zzeekk/spark/temporalquery/IntervalDef.scala

```diff
@@ -0,0 +1,196 @@
+   */
+  def predecessor(value: T): T
+  /**
+   * Method get the successor for a scala value of type T for this interval axis definition
+   */
+  def successor(value: T): T
+
+  /**
+   * Expression to floor a column
+   */
+  def getFloorExpr(valueCol: Column): Column = {
+    val udfTransform = udf((v: Any) => Option(v).map(x => floor(x.asInstanceOf[T])))
+    udfTransform(valueCol)
+  }
+  /**
+   * Expression to ceil a column
+   */
+  def getCeilExpr(valueCol: Column): Column = {
+    val udfTransform = udf((v: Any) => Option(v).map(x => ceil(x.asInstanceOf[T])))
+    udfTransform(valueCol)
+  }
+  /**
+   * Expression get predecessor for a column
+   */
+  def getPredecessorExpr(valueCol: Column): Column = {
+    val udfTransform = udf((v: Any) => Option(v).map(x => predecessor(x.asInstanceOf[T])))
+    udfTransform(valueCol)
+  }
+  /**
+   * Expression get successor for a column
+   */
+  def getSuccessorExpr(valueCol: Column): Column = {
+    val udfTransform = udf((v: Any) => Option(v).map(x => successor(x.asInstanceOf[T])))
+    udfTransform(valueCol)
+  }
+  /**
+   * make sure value is between min- and maxValue
+   */
+  @inline def fitToBoundaries(value: T): T = least(greatest(value, minValue), maxValue)
+  def getFitToBoundariesExpr(valueCol: Column): Column = {
+    functions.when(valueCol.isNotNull, functions.least(functions.greatest(valueCol, functions.lit(minValue)), functions.lit(maxValue)))
+  }
+
+  // Helpers
+  @inline private def least(values: T*): T = values.min
+  @inline private def greatest(values: T*): T = values.max
+}
+
+/**
+ * A closed interval is an interval which includes its lower and upper bound.
+ * Use this for discret interval axis.
+ * @param minValue minimum value of the interval axis
+ * @param maxValue maximum value of the interval axis
+ * @tparam T: scala type for discrete interval axis, e.g. Timestamp, Integer, ...
+ */
+case class ClosedInterval[T: Ordering: TypeTag](override val minValue: T, override val maxValue: T, discreteAxisDef: DiscreteAxisDef[T]) extends IntervalDef[T] {
+  assert(minValue == floor(minValue), s"minValue $minValue is not discrete value of the axis")
+  assert(maxValue == floor(maxValue), s"minValue $maxValue is not discrete value of the axis")
+  override def isInIntervalExpr(valueCol: Column, fromCol: Column, toCol: Column): Column = {
+    fromCol <= valueCol && valueCol <= toCol
+  }
+  def isValidIntervalExpr(fromCol: Column, toCol: Column): Column = {
+    fromCol <= toCol
+  }
+  override def intervalJoinExpr(fromCol1: Column, toCol1: Column, fromCol2: Column, toCol2: Column): Column = {
+    fromCol1 <= toCol2 and toCol1 >= fromCol2
+  }
+  def floor(value: T): T = fitToBoundaries(discreteAxisDef.floor(value))
+  def ceil(value: T): T = fitToBoundaries(discreteAxisDef.ceil(value))
+  def predecessor(value: T): T =
+    fitToBoundaries(if (value==maxValue) value else discreteAxisDef.predecessor(value)) // max value has no predecessor
```

Comment:
```suggestion
    fitToBoundaries(if (value==minValue) value else discreteAxisDef.predecessor(value)) // min value has no predecessor
```

Reply:
No, thats actually right i think. Unit tests fail otherwise, e.g. temporalCleanupExtend_dfLeft:
```
   Actual 
  actual.count() =  3
struct<id:int,wert_l:double,_defined:boolean,gueltig_ab:timestamp,gueltig_bis:timestamp>
+---+------+--------+-------------------+-----------------------+
|id |wert_l|_defined|gueltig_ab         |gueltig_bis            |
+---+------+--------+-------------------+-----------------------+
|0  |null  |false   |0001-01-01 00:00:00|2017-12-09 23:59:59.999|
|0  |4.2   |true    |2017-12-10 00:00:00|2018-12-08 23:59:59.999|
|0  |null  |false   |2018-12-09 00:00:00|9999-12-30 23:59:59.999|
+---+------+--------+-------------------+-----------------------+

   Expected 
  expected.count() =  3
struct<id:int,wert_l:double,_defined:boolean,gueltig_ab:timestamp,gueltig_bis:timestamp>
+---+------+--------+-------------------+-----------------------+
|id |wert_l|_defined|gueltig_ab         |gueltig_bis            |
+---+------+--------+-------------------+-----------------------+
|0  |null  |false   |0001-01-01 00:00:00|2017-12-09 23:59:59.999|
|0  |4.2   |true    |2017-12-10 00:00:00|2018-12-08 23:59:59.999|
|0  |null  |false   |2018-12-09 00:00:00|9999-12-31 00:00:00    |
+---+------+--------+-------------------+-----------------------+

```

Reply:
Hi Zack

I think there is a confusion between upper/lower bound of a specific interval and min/max value of the time axis:
- upper/lower bound are the bounds of a specific interval, e.g. [0.0, 1.0] has as lower bound 0.0 and 1.0 as upper bound.
- min/max value are values which designate -∞/+∞.

Of course -∞/+∞ do not have predecessor nor sucessor, but upper/lower bound do so except they coincide with -∞/+∞

I am a bit confused about
```
case class TemporalClosedIntervalQueryConfig(...
    val minDate: Timestamp = intervalDef.lowerBound
    val maxDate: Timestamp = intervalDef.upperBound
```
Maybe I misunderstood your intention: By defining an interval, e.g. a `ClosedInterval`, do you mean defining a specific interval or the value axis?

Maybe we should discuss this in a meeting.

Reply:
Discussed and IntervalDef.lower/upperBound renamed to lower/upperHorizon.